From 9fab6c4c8fd5f7dff89d60f4a96c1ec810a03fae Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Fri, 17 May 2013 11:48:17 +0900
Subject: [PATCH] ARM: shmobile: r8a7778: add a register region for an ICB
 errata workaround to UIO

This patch allows to access a register to control ICB cache for
an ICB errata workaround from userland. With the register, the ICB
cache can be set to being enabled or disabled, and flushed as required.
---
 arch/arm/mach-shmobile/setup-r8a7778.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/arch/arm/mach-shmobile/setup-r8a7778.c b/arch/arm/mach-shmobile/setup-r8a7778.c
index 0a72e86..818fe40 100644
--- a/arch/arm/mach-shmobile/setup-r8a7778.c
+++ b/arch/arm/mach-shmobile/setup-r8a7778.c
@@ -465,6 +465,32 @@ static struct platform_device vpu_device = {
 	.num_resources  = ARRAY_SIZE(vpu_resources),
 };
 
+/* ICB Cache Control */
+static struct uio_info icb_cache_ctrl_platform_data = {
+	.name = "ICB_CACHE_CTRL",
+	.version = "0",
+	.irq = -1,
+};
+
+static struct resource icb_cache_ctrl_resources[] = {
+	[0] = {
+		.name   = "ICB_CACHE_CTRL",
+		.start  = 0xfd050000,
+		.end    = 0xfd050097,
+		.flags  = IORESOURCE_MEM,
+	},
+};
+
+static struct platform_device icb_cache_ctrl_device = {
+	.name           = "uio_pdrv_genirq",
+	.id             = 3,
+	.dev = {
+		.platform_data  = &icb_cache_ctrl_platform_data,
+	},
+	.resource       = icb_cache_ctrl_resources,
+	.num_resources  = ARRAY_SIZE(icb_cache_ctrl_resources),
+};
+
 static struct platform_device *r8a7778_early_devices[] __initdata = {
 	&scif0_device,
 	&scif1_device,
@@ -483,6 +509,7 @@ static struct platform_device *r8a7778_early_devices[] __initdata = {
 	&meram_device,
 	&vio_device,
 	&vpu_device,
+	&icb_cache_ctrl_device,
 };
 
 static struct platform_device *r8a7778_late_devices[] __initdata = {
-- 
1.8.1.2

